### PR TITLE
[1.26] kubelet: devices: skip allocation for running pods #118635 

### LINKED ISSUE #118559: Running pods with devices are terminated if kubelet is restarted
### What happened?

In KubeVirt project, we now see a regression when running on Kubernetes `1.25.10 | 1.26.5 | 1.27.2`. If kubelet is restarted on a node, then all the existing and running workloads that use devices are terminated with `UnexpectedAdmissionError`:
```
Warning  UnexpectedAdmissionError  45s   kubelet            Allocate failed due to no healthy devices present; cannot allocate unhealthy devices devices.kubevirt.io/kvm, which is unexpected
Normal   Killing                   42s   kubelet            Stopping container compute
```
KubeVirt runs virtual machines inside pods and uses a device plugin to advertise e.g. `/dev/kvm` on the nodes.

Presumably, this PR changed the behavior: https://github.com/kubernetes/kubernetes/pull/116376
Original issue: https://github.com/kubernetes/kubernetes/issues/109595

### What did you expect to happen?

A potential restart of kubelet should not interrupt the running workloads.

### How can we reproduce it (as minimally and precisely as possible)?
with KubeVirt:

- run a KubeVirt VM
- `pkill kubelet`
- observe that the workload pod gets terminated

or with https://github.com/k8stopologyawareschedwg/sample-device-plugin

- make deploy
- make test-both
- pkill kubelet
- the pod gets restarted

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

This affects the 1.25.x, 1.26.x and 1.27.x branches.

1.25.10 | 1.26.5 | 1.27.2

</details>

### Cloud provider

<details>

N/A

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by ffromani**
#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:
Cherry-pick of #118635 to branch `release-1.26` through #119432. Cherry pick per se done using `hack/cherry_pick_pull.sh`

#### Original description
When kubelet initializes, runs admission for pods and possibly allocate requested resources. We need to distinguish between node reboot (no containers running) versus kubelet restart (containers potentially running).

Running pods should always survive kubelet restart. This means that device allocation on admission should not be attempted, because if a container requires devices and is still running when kubelet is restarting, that container already has devices allocated and working.

Thus, we need to properly detect this scenario in the allocation step and handle it explicitely. We need to inform the devicemanager about which pods are already running.

#### Which issue(s) this PR fixes:
Fixes #118559

#### Special notes for your reviewer:
Implements the first approach [proposed](https://github.com/kubernetes/kubernetes/issues/118559#issuecomment-1584636260) in [the thread](https://github.com/kubernetes/kubernetes/issues/118559#issuecomment-1584802420), so we make the devicemanager treat running pod differently.

This approach was chosen because it seems simpler to make self-contained and easier to backport.

The devicemanager already tracks (with the help of the checkpoint files) which containers got devices assigned to them, which by definition means these containers passed its admission. The missing bit is safely learning which container are already running when initializing, and for that we extend the [existing `buildContainerMapFromRuntime`](https://github.com/kubernetes/kubernetes/blob/v1.28.0-alpha.2/pkg/kubelet/cm/container_manager_linux.go#L567)

#### Does this PR introduce a user-facing change?
```release-note
Fixes regression in 1.26.5 causing running pods with devices to be terminated if kubelet is restarted
```

### CODE REVIEW

**Comment by ffromani** [comment]
/retest

seems unrelated

**Comment by ffromani** [comment]
/retest

looks infra issue on azure? Will file an issue if it is the case

**Comment by haircommander** [comment]
LGTM

/hold

for https://github.com/kubernetes/kubernetes/pull/119432

**Comment by ffromani** [comment]
rebased to fix a conflict

**Review by swatisehgal — COMMENTED** [review]
/lgtm

Re-applying as the label was removed due to rebase

**Review by mrunalp — APPROVED** [review]

**Comment by bobbypage** [comment]
/cc @kubernetes/release-managers 

Release managers, can you please take a look at this cherrypick? Thank you!

**Comment by ffromani** [comment]
/hold cancel

1.27 woes solved - and I think we won't merge cherry-picks out of expected order anyway

**Review by xmudrii — APPROVED** [review]
For RelEng:
/lgtm
/approve

**Pull request merged**
